File: codevillage/codevillage-170.py
mod_num = 10 ** 9 + 7


def main():
    global mod_num
    n, m = tuple(map(int, input().split()))
    if m <= 2 or n <= 0:
        print(0)
        return
    result = m * (m - 1) * (m - 2)
    mul_num = (m - 1) * (m - 2) * max(m - 3, 1)
    for i in range(n - 1):
        result = result % mod_num
        result *= mul_num
    result = result % mod_num
    print(result)


if __name__ == '__main__':
    main()

File: codevillage/codevillage-170.cpp
#include <iostream>
using namespace std;
typedef long long ll;

ll n, m, result = 1, mod = 1000000007, f;

int main(){
    cin >> n >> m;
    if (m <= 2){
        cout << 0;
        return 0;
    }
    f = (m - 1) * (m - 2) * max(m - 3, (ll) 1);
    for (int i = 0; i < n; i ++){
        if (i == 0){
            result = result * m * (m - 1) * (m - 2) % mod;
        }
        else{
            result = result * f % mod;
        }
    }
    cout << result;
}